Bug 1133752 - Convert Mozmill software-update.js library to Python/Marionette

bobsilverberg opened the pull request:
@whimboo This is not yet complete, and I'm not sure how good it is, as I was mostly hacking my way through it, but everything (except `active_update`) seems to be working.

I'm sure both the code and tests need quite a bit of work, but it's a start. I'd appreciate it if you could take a look at it when you have time and provide any feedback you may have.

bobsilverberg commented:
@whimboo This is now complete, in that all of the methods have been converted. I would not be surprised if there are some problems with `init_update_tests` and `assert_update_applied` as I'm not really sure how info is going to get populated and passed around, but hopefully this PR is a good start.

I'm going to turn my attention back to about-window.js now.

whimboo commented on firefox_puppeteer/__init__.py:
```diff
@@ -79,6 +79,14 @@ def security(self):
+        Provides support for an easy access to the update process.
+
+        See the :class:`~api.software_update.SoftwareUpdate` reference.
+        """
```
I think that we don't need to expose the software update module to Puppeteer. We only have 2 tests at all using it. They can import the necessary bits when necessary.

whimboo replied:
Please make sure to add the appropriate sphinx files so that the module gets documented.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+
+    REGEX_UPDATE_CHANNEL_PREF = re.compile(r'("app\.update\.channel", ")([^"].*)(?=")')
+
+    def __init__(self, software_update):
```
This class should be based on BaseLib instead of object, so it would require the marionette getter as parameter to the __init__ method.

whimboo replied:
Also to add, this class needs nothing from the software_update class and should stay independent from it. If you have to pass data lets get a property added.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        self.prefs = software_update.prefs
+
+        self.ini_file_path = self.marionette.execute_script("""
+            Cu.import('resource://gre/modules/Services.jsm');
```
nit: For JS code we have an indentation of 2 chars.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+    @property
+    def config(self):
+        config = ConfigParser.ConfigParser()
+        config.readfp(open(self.ini_file_path))
```
What happens in case of a failure in readfp? Is the file getting closed? I think it would be better to work with `with` here, or?

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+            """)
+
+    @property
+    def config(self):
```
This property is for internal use. I would prefix its name with an underscore.

whimboo replied:
And may name it `_content`?

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        return self.ini_file_path
+
+    @property
+    def config_file_contents(self):
```
This property is also for internal use only. There is no need that we have to expose it. Lets mark it as private.

bobsilverberg replied:
This is actually used in `init_update_tests`, which is in `SoftwareUpdate` so I think it needs to remain public.

whimboo replied:
Lets keep it for now until we can make use of a Pythonic INI reader and writer. We can refactor this out later.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        channels = self.config.get(self.INI_SECTION, self.INI_OPTION)
+        return channels.split(',')
+
+    def set_channels(self, channels):
```
Make this a setter property instead. It simplifies the usage of this class.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        return channels.split(',')
+
+    def set_channels(self, channels):
+        channels = ','.join(channels)
```
I think you can directly put this into the call to `set`.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        new_config = self.config
+        new_config.set(self.INI_SECTION, self.INI_OPTION, channels)
+        with open(self.ini_file_path, 'wb') as f:
+            new_config.write(f)
```
Have you checked that this produces the correct output? When I was implementing this class originally I have seen that the Python config library added blanks around the `=` sign for this key. This will break reading the ini file in Firefox.

bobsilverberg replied:
You are correct. My tests didn't catch that. I guess we cannot use `ConfigParser`, which is too bad. :(

whimboo replied:
> You are correct. My tests didn't catch that. I guess we cannot use
> |ConfigParser|, which is too bad. :(

Please file a new follow-up bug for that. Maybe we can find a hacker who
wants to take it.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        new_config = self.config
+        new_config.set(self.INI_SECTION, self.INI_OPTION, new_channels)
+        with open(self.ini_file_path, 'wb') as f:
+            new_config.write(f)
```
This is duplicating the code from `set_channels`. Please call this method with the final list of channels instead.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        current_channels = self.channels
+        for channel in channels:
+            if channel in current_channels:
+                current_channels.remove(channel)
```
Have you done list comprehensions yet? It's a very powerful future we can use here.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        new_config = self.config
+        new_config.set(self.INI_SECTION, self.INI_OPTION, new_channels)
+        with open(self.ini_file_path, 'wb') as f:
+            new_config.write(f)
```
Same as above. Please call `set_channels` instead of duplicating the code.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+
+    REGEX_UPDATE_CHANNEL_PREF = re.compile(r'("app\.update\.channel", ")([^"].*)(?=")')
+
+    def __init__(self, software_update):
```
Same as for the other class. Please make BaseLib the base class.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        self.marionette = software_update.marionette
+        self.prefs = software_update.prefs
+
+        self.config_file_path = self.marionette.execute_script("""
```
Maybe just `file_path`? It's not really a config file.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+            """)
+
+    @property
+    def config_file_contents(self):
```
Maybe just `_content`?

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+
+        :returns: Current default update channel
+        """
+        matches_list = re.findall(self.REGEX_UPDATE_CHANNEL_PREF, self.config_file_contents)
```
We only have a single entry here. What would be the benefit of `findall` compared to `search`?

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        matches_list = re.findall(self.REGEX_UPDATE_CHANNEL_PREF, self.config_file_contents)
+        assert len(matches_list) == 1, 'Update channel value has been found'
+        matches_tuple = matches_list[0]
+        assert len(matches_tuple) == 2, 'Update channel value has been found'
```
Additionally to the above question I think `search` would make it easier here.

bobsilverberg replied:
Good point, `search` is better.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+
+        return matches_tuple[1]
+
+    def set_default_channel(self, channel):
```
Please make it also a setter property.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+    """Class to handle the update channel as listed in channel-prefs.js
+    """
+
+    REGEX_UPDATE_CHANNEL_PREF = re.compile(r'("app\.update\.channel", ")([^"].*)(?=")')
```
nit: It's not a pref but a pref_name. But I would simply drop this suffix.

whimboo commented on firefox_puppeteer/tests/test_software_update.py:
```diff
@@ -0,0 +1,88 @@
+        self.assertIn('Firefox', self.software_update.update_url(True))
+        self.assertIn('force=1', self.software_update.update_url(True))
+
+    def test_update_channel_channel(self):
```
Each class in the software_update.py file should get it's own test class in this file.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        self._aus = self.marionette.execute_script("""
+            return Cc['@mozilla.org/updates/update-service;1']
+                .getService(Ci.nsIApplicationUpdateService);
+            """)
```
Lets not return the service here. It will be JSON string but no real service anymore, and you won't be able to convert it back. We can access properties from it but no methods. So lets simply add this code to all the methods where it is used.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        if mozinfo.isMac:
+            abi += self.marionette.execute_script("""
+                let macutils = Cc['@mozilla.org/xpcom/mac-utils;1']
+                    .getService(Ci.nsIMacUtils);
```
For a continuation line the dot should be aligned with the `C`.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+                let macutils = Cc['@mozilla.org/xpcom/mac-utils;1']
+                    .getService(Ci.nsIMacUtils);
+                if (macutils.isUniversalBinary)
+                    return '-u-' + macutils.architecturesInBinary;
```
Please always use brackets for blocks in JS.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+
+        self.update_channel = UpdateChannel(self)
+        self.mar_channels = MARChannels(self)
+        self.active_update = ActiveUpdate(self)
```
Those would have to appear in the sphinx docs. Not sure how you would do that. I propose you assign the instances to private variables and create a property for each of those.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        :returns: Status if the user has the permissions
+        """
+
+        return self._aus['canCheckForUpdates'] and self._aus['canApplyUpdates']
```
As mentioned above lets work on the real aus service object.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+            return  Services.vc.compare(arguments[0], arguments[1]);
+        """, script_args=[info['build_post']['version'], info['build_pre']['version']])
+
+        assert check >= 0, 'The version of the upgraded build is higher or equal'
```
Maybe it is better to use assert methods from the unittest class here.

bobsilverberg replied:
That is not what is currently being done in `prefs` (e.g., [1])  or `windows` [2]. I can make that change if you wish, but I wanted to check with you first for consistency.

[1] https://github.com/mozilla/firefox-ui-tests/blob/master/firefox_puppeteer/api/prefs.py#L153
[2] https://github.com/mozilla/firefox-ui-tests/blob/master/firefox_puppeteer/ui/windows.py#L318

whimboo replied:
>> +        assert check >= 0, 'The version of the upgraded build is higher or equal'
> 
> That is not what is currently being done in |prefs| (e.g., [1]) or
> |windows| [2]. I can make that change if you wish, but I wanted to check
> with you first for consistency.
> 
> [1]
> https://github.com/mozilla/firefox-ui-tests/blob/master/firefox_puppeteer/api/prefs.py#L153
> [2]
> https://github.com/mozilla/firefox-ui-tests/blob/master/firefox_puppeteer/ui/windows.py#L318

In those other libs we use `assert` for checking parameters. But I see
your point. @chmanchester what do you think? Thinking more about it, it
may not a good idea to base library modules on the unittest module.

chmanchester replied:
This is somewhere between a library and a test, I think either would be acceptable.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        return {
+            'buildid': self.app_info.appBuildID,
+            'channel': self.update_channel.channel,
+            'disabled_addons': self.prefs.get_pref('PREF_DISABLED_ADDONS'),
```
The parameter in `get_pref` please without the quotes. :)

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+
+        return self._aus['canCheckForUpdates'] and self._aus['canApplyUpdates']
+
+    def assert_update_applied(self, update_data):
```
I would suggest we also keep the following order in a class: properties, and then methods. If it's not such a hassle to move some parts around here.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+            let update_index = arguments[0];
+            return ums.getUpdateAt(update_index);
+
+        """, script_args=[update_index])
```
Same as for the aus service above. We cannot return a JS object. It will be a JSON string and cannot be used anymore. :S

bobsilverberg replied:
It turns out I never need to call this `get_update_at` anyway. I think I added it while testing some things, so I will just remove the entire method.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+            persisted['update']['default_update_channel'] = {
+                'content': self.update_channel.config_file_contents,
+                'path': self.update_channel.config_file_path}
+            self.update_channel.set_default_channel(persisted['update']['channel'])
```
Given that the UpdateChannel class lives in Python now we won't have to store the defaults anymore, but simply ensure to restore the file with its original content at the end of the test. Maybe we can call this method `setUp` and add a new `tearDown` method to do that?

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+            # Backup the original content and the path of the update-settings.ini file
+            persisted['update']['default_mar_channels'] = {
+                'content': self.mar_channels.config_file_contents,
+                'path': self.mar_channels.config_file_path}
```
Same here. We can directly restore the content at the end of the test.

bobsilverberg replied:
I'm sorry, @whimboo, but I cannot tell what the context of this comment was. I'm not sure what the "Same here." refers to. Can you be more explicit about what needs to change in here? I'm still not totally clear about what we're planning to do to get rid of `persisted` and if it means we need more major changes in here.

whimboo replied:
Good question! I assume it was also related to restoring the content of the file. With the new bug to be filed we can cover both cases. So just file the bug as mentioned above for the update-settings.ini file.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        assert patch_count, 'An active update has been found'
+
+        assert patch_count == 1 or patch_count == 2,\
+            'An update must have one or two patches included'
```
Can we use the `unittest` equivalent methods here too?

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        # Throw when isCompleteUpdate is called without an update. This should
+        # never happen except if the test is incorrectly written.
+        patch_count = self.active_update.patch_count
+        assert patch_count, 'An active update has been found'
```
You have added `active_update.exists` so I think it is what we should use here.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+            let aus = Cc['@mozilla.org/updates/update-service;1']
+                .getService(Ci.nsIApplicationUpdateService);
+            var dir = aus.getUpdatesDirectory();
+            return dir.path;
```
I would directly do a `return aus.getUpdatesDirectory().path` here.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        """Returns the type of the active update."""
+        return self.active_update.type
+
+    def update_url(self, force=False):
```
This is not a property, so please add the `get_` prefix to the method name.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        url = self.prefs.get_pref(self.PREF_APP_UPDATE_URL, '')
+        # get the next two prefs from the default branch
+        dist = self.prefs.get_pref(self.PREF_APP_DISTRIBUTION, True) or 'default'
+        dist_version = self.prefs.get_pref(self.PREF_APP_DISTRIBUTION_VERSION, True) or 'default'
```
Move those 3 lines after the following if block, so we don't run this code if url is empty.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+            return None
+
+        # Not all placeholders are getting replaced correctly by formatURL
+        url = re.sub(r'%PRODUCT%', self.app_info.name, url)
```
I would suggest we keep using the string replace method here, which is much faster than any regex.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        url = self.marionette.execute_script("""
+            Cu.import("resource://gre/modules/Services.jsm");
+
+            var url = arguments[0];
```
Please lets use `let` in all the places.

whimboo replied:
In this case we could even get rid of this line.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+        return url
+
+
+class ActiveUpdate(object):
```
Same here for BaseLib as super class.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+            return active_update[arguments[0]];
+        """, script_args=[attr])
+
+        if value is not None:
```
`if value` should be enough.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+                .getService(Ci.nsIUpdateManager);
+            let active_update = ums.activeUpdate;
+            let patch_index = arguments[0];
+            return active_update.getPatchAt(patch_index);
```
I checked DXR and so far we really only use properties so we should be fine in returning the JSON data for the interface.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+            let ums = Cc['@mozilla.org/updates/update-manager;1']
+                .getService(Ci.nsIUpdateManager);
+            let active_update = ums.activeUpdate;
+            return active_update.patchCount;
```
Let directly return without the extra variable.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,544 @@
+            let ums = Cc['@mozilla.org/updates/update-manager;1']
+                .getService(Ci.nsIUpdateManager);
+            let active_update = ums.activeUpdate;
+            return active_update.selectedPatch.finalURL;
```
Same here. return directly.

whimboo replied:
Hm, maybe we can also have a `selected_patch` property only and return the JSON object? That way we have only a single method and using that property will become easier.

bobsilverberg commented:
@whimboo I have added a commit that addresses the comments for MarChannels. I plan on adding a commit for each class, as I address them, You may want to wait for all of my updates before re-reviewing. I will add a comment when they are all complete.

bobsilverberg commented:
@whimboo I have addressed all of your review comments, except for the one about `init_update_tests` as I have to admit I'm not really sure what is going on in that method and what actually needs to be done. If you can re-review, then maybe we can have a short meeting to discuss what's left, including `init_update_tests`.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+        """)
+
+
+class MARChannels(BaseLib):
```
Mind moving this class up so we have an alphabetical sorting?

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+        self.ini_file_path = self.marionette.execute_script("""
+          Cu.import('resource://gre/modules/Services.jsm');
+
+          let file = Services.dirsvc.get('GreD', Ci.nsIFile);
```
I would advise to not use the shortcuts for `Components.*` here. It will fail if this code is executed from a window which does not have those defined. Please change that across the file.

bobsilverberg replied:
I'm not entirely sure what you mean. Do you mean replace `Ci.` with `Components.interfaces.`? Do I also need an import for that before calling it?

whimboo replied:
Correct. `Ci.` will only work inside a browser window but will most likely fail if used in other chrome window types. Given that you are in chrome scope no import or such things are necessary.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+    INI_OPTION = 'ACCEPTED_MAR_CHANNEL_IDS'
+
+    def __init__(self, marionette_getter):
+        BaseLib.__init__(self, marionette_getter)
```
Please add a docstring here.

bobsilverberg replied:
A docstring for `__init__`? None of the other classes have docstrings for their `__init__` method. I would think that would be somewhat superfluous. Or do you mean a comment to describe the code below, that sets the path for the ini file?

whimboo replied:
You are right. We should do a follow-up bug for all those cases. No need to do it now.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+
+    @property
+    def config_file_path(self):
+        return self.ini_file_path
```
Same here and other places for docstrings.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+                           .getService(Ci.nsIINIParserFactory);
+
+          let file = Services.dirsvc.get('GreD', Ci.nsIFile);
+          file.append('update-settings.ini');
```
We have the path as `self.ini_file_path` so please push this in here as a script_arg.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+        :param channels: a comma delimited string of channel names
+        """
+
+        self.marionette.execute_script("""
```
nit: There is not a blank line after a docstring.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+                           .getService(Ci.nsIINIParserFactory);
+
+          let file = Services.dirsvc.get('GreD', Ci.nsIFile);
+          file.append('update-settings.ini');
```
Same here as above for the file path.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+          """, script_args=[self.INI_SECTION, self.INI_OPTION, channels])
+
+    def add_channels(self, channels):
+        new_channels = ','.join(self.channels) + ',' + channels
```
This would add duplicates to the list. I would suggest that we pass in an array of channels here and not a string, so we can work with a Set().

whimboo replied:
Also we should do the stringification inside the channels getter and setter.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+    def remove_channels(self, channels):
+        current_channels = self.channels
+        new_channels = [channel for channel in current_channels if channel not in channels]
+        new_channels = ','.join(new_channels)
```
Same here as for `add_channels`.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+
+    def remove_channels(self, channels):
+        current_channels = self.channels
+        new_channels = [channel for channel in current_channels if channel not in channels]
```
When looking at this it reminds me of a comment from @chmanchester and that we should better use `Set() - Set()` here. Maybe we can hold `self.channels` as a set on that class directly.

whimboo commented on firefox_puppeteer/docs/api/software_update.rst:
```diff
@@ -0,0 +1,27 @@
+MARChannels
+-----------
+
+.. autoclass:: ActiveUpdate
```
c&p failure.

whimboo commented on firefox_puppeteer/docs/api/software_update.rst:
```diff
@@ -0,0 +1,27 @@
+UpdateChannel
+-------------
+
+.. autoclass:: ActiveUpdate
```
c&p failure.

whimboo commented on firefox_puppeteer/docs/api/software_update.rst:
```diff
@@ -0,0 +1,27 @@
+ActiveUpdate
+------------
+
+.. autoclass:: ActiveUpdate
```
c&p failure.

whimboo commented on firefox_puppeteer/tests/test_software_update.py:
```diff
@@ -0,0 +1,108 @@
+        self.software_update.force_fallback()
+        with open(os.path.join(self.software_update.staging_directory, 'update.status'), 'r') as f:
+            content = f.read()
+        self.assertEqual(content, 'failed: 6\n')
```
To prevent follow-up failures you might want to remove this file in tearDown or maybe better have a try/finally here.

whimboo commented on firefox_puppeteer/tests/test_software_update.py:
```diff
@@ -0,0 +1,108 @@
+    def test_get_update_url(self):
+        self.assertIn('Firefox', self.software_update.get_update_url())
+        self.assertIn('Firefox', self.software_update.get_update_url(True))
+        self.assertIn('force=1', self.software_update.get_update_url(True))
```
Maybe caching the call to `get_update_url()`? Should we also add `assertNotIn()` for `force=1` if not forced?

whimboo commented on firefox_puppeteer/tests/test_software_update.py:
```diff
@@ -0,0 +1,108 @@
+        FirefoxTestCase.setUp(self)
+        self.software_update = SoftwareUpdate(self.get_marionette)
+        self.saved_mar_channels = self.software_update.mar_channels.channels
+        self.software_update.mar_channels.channels = 'expected,channels'
```
MAR channel logic is not needed in this test class.

bobsilverberg replied:
I included it because `build_info` includes the MARChannels, so I wanted to be able to set an expected value for that in order to do the assert. I can remove it and just use `assertTrue` if you would prefer that.

whimboo replied:
> I included it because |build_info| includes the MARChannels, so I wanted
> to be able to set an expected value for that in order to do the assert.
> I can remove it and just use |assertTrue| if you would prefer that.

I see. That makes sense then. So lets leave it as is. Thanks.

whimboo commented on firefox_puppeteer/tests/test_software_update.py:
```diff
@@ -0,0 +1,108 @@
+
+    def tearDown(self):
+        self.software_update.update_channel.default_channel = self.saved_channel
+        FirefoxTestCase.tearDown(self)
```
I would suggest a try/finally here so we can ensure that the super `tearDown` call gets executed.

whimboo commented on firefox_puppeteer/tests/test_software_update.py:
```diff
@@ -0,0 +1,108 @@
+
+    def test_update_channel_set_default_channel(self):
+        self.software_update.update_channel.default_channel = 'new_channel'
+        self.assertEqual('new_channel', self.software_update.update_channel.default_channel)
```
Please flip so that the reference comes as second argument.

whimboo commented on firefox_puppeteer/tests/test_software_update.py:
```diff
@@ -0,0 +1,108 @@
+
+    def tearDown(self):
+        self.software_update.mar_channels.channels = self.saved_mar_channels
+        FirefoxTestCase.tearDown(self)
```
Same for try/finally here.

whimboo commented on firefox_puppeteer/tests/test_software_update.py:
```diff
@@ -0,0 +1,108 @@
+        FirefoxTestCase.tearDown(self)
+
+    def test_mar_channels_channels(self):
+        self.assertEqual(['expected', 'channels'], self.software_update.mar_channels.channels)
```
Please flip for those tests too.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+        :returns: ABI version
+        """
+
+        abi = self.app_info.XPCOMABI
```
nit: no blank line after a docstring. Please also check other instances.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+
+        self.update_channel = UpdateChannel(marionette_getter)
+        self.mar_channels = MARChannels(marionette_getter)
+        self.active_update = ActiveUpdate(marionette_getter)
```
Those properties would not appear in the sphinx docs. Maybe we can tag them with a special flag?

bobsilverberg replied:
I could add properties for them, but that seems a bit redundant. Do you know how to flag them to be picked up for the docs, or is that something you want me to figure out?

whimboo replied:
I don't know about the details, but I would propose we make the instances private by adding a _ prefix to them. Then only add a getter property for each of them. That way we can ensure that no outer code can mess up with those instances.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+
+        aus_dict = self.marionette.execute_script("""
+          return Cc['@mozilla.org/updates/update-service;1']
+            .getService(Ci.nsIApplicationUpdateService);
```
nit: please align the dot with `C`

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+            .getService(Ci.nsIApplicationUpdateService);
+          """)
+
+        return aus_dict['canCheckForUpdates'] and aus_dict['canApplyUpdates']
```
Maybe better to directly do the check in JS and return a boolean via `execute_script`?

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+    def staging_directory(self):
+        return self.marionette.execute_script("""
+          let aus = Cc['@mozilla.org/updates/update-service;1']
+                   .getService(Ci.nsIApplicationUpdateService);
```
nit: indentation.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+        if persisted['update']['channel']:
+            # Backup the original content and the path of the channel-prefs.js file
+            persisted['update']['default_update_channel'] = {
+                'content': self.update_channel.file_contents,
```
As we talked in the meeting yesterday please file a follow-up bug so we do copies of the files instead.

bobsilverberg replied:
Done. https://bugzilla.mozilla.org/show_bug.cgi?id=1138946

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,491 @@
+
+          return Services.urlFormatter.formatURL(arguments[0]);
+
+          """, script_args=[url])
```
nit: indentation issue. Align it with `u` from url. Maybe drop the blank lines in the JS code.

whimboo commented:
I made a couple of comments for the current code. It all looks fine code-wise in nearly all cases. When working on the next commit please also rebase against master. Looks like there are merge conflicts.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,499 @@
+
+    def __getattr__(self, attr):
+        value = self.marionette.execute_script("""
+          let ums = Cc['@mozilla.org/updates/update-manager;1']
```
You still have `Cc` in your patch which would also break us. Please also use `Components.classes` everywhere.

bobsilverberg replied:
Sorry. This is the first time I've ever worked with any of this Firefox API stuff, so I'm not really sure what I'm doing. Hopefully `Cc` is the last one that needs to be changed.

whimboo replied:
> Sorry. This is the first time I've ever worked with any of this Firefox
> API stuff, so I'm not really sure what I'm doing. Hopefully |Cc| is the
> last one that needs to be changed.

No worries Bob! You are doing great on all that. Yes, `Cc` was the last
shotcut which needed to be ripped off. Thanks.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,499 @@
+
+    @property
+    def exists(self):
+        """Is there an active update?
```
nit: We don't use questions everywhere else. So just `Checks if there is an active update`.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,499 @@
+    def __init__(self, marionette_getter):
+        BaseLib.__init__(self, marionette_getter)
+
+        self.ini_file_path = self.marionette.execute_script("""
```
With the `config_file_path` property on this class we should make this property private.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,499 @@
+    def channels(self, channels):
+        """Set the channels in the update-settings.ini file.
+
+        :param channels: an array of channel names
```
nit: Start the description with a capital letter.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,499 @@
+        :param channels: an array of channel names to add
+        """
+        new_channels = set(self.channels) | set(channels)
+        self.channels = new_channels
```
As mentioned in my last review cycle you can make `self.channels` to be a Set by default. So `channels` can return a Set instead of an array. That would make it easier across the file to use it. Maybe also assign it directly?

whimboo commented on firefox_puppeteer/tests/test_software_update.py:
```diff
@@ -0,0 +1,122 @@
+            with open(
+                    os.path.join(
+                            self.software_update.staging_directory, 'update.status'), 'r')as f:
+                content = f.read()
```
The indentation of the lines above look a bit strange but for a test it will be fine.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,499 @@
+
+    @property
+    def mar_channels(self):
+        """ Holds a reference to a MARChannels object."""
```
nit: you may want to linkify the class name like ":class:MARChannels". This might also apply to some other methods.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,499 @@
+          return aus.canCheckForUpdates && aus.canApplyUpdates;
+        """)
+
+        return bool(allowed)
```
`execute_script` already returns a boolean value, so you can directly return it.

whimboo commented:
I went through the code again by only watching those areas I have commented on earlier, and you applied fixes. I hope you didn't change something else. If you did please let me know, otherwise I have to do a full review again. 

Most of the remaining changes can quickly be done, and as you see the number of comments decreases rapidly. So I think that we can get it merged by tomorrow. Thanks a lot Bob!

bobsilverberg commented:
@whimboo I have only been making changes specified by your comments, so that should be fine. Also, I always ensure to make all of the changes for a review cycle in a single commit, so you can always look at the diff for the commit to see exactly what has changed since the last review.

I have now addressed the most recent set of comments and pushed a new commit.

whimboo commented on firefox_puppeteer/api/software_update.py:
```diff
@@ -0,0 +1,496 @@
+
+        :param channels: A set of channel names to remove
+        """
+        self.channels = self.channels - channels
```
Much cleaner and with more performance!  This looks way better.

whimboo commented:
This PR has been merged as https://github.com/mozilla/firefox-ui-tests/commit/2c3c1df786fe8843fb89c3c85f3eab7e8ecb73cc
